pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -0,0 +1,223 @@
+	watcherapi "k8s.io/kubernetes/pkg/kubelet/apis/pluginregistration/v1alpha1"
+	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
+)
+
+// RegisterCallbackFn is the type of the callback function that handlers will provide
+type RegisterCallbackFn func(pluginName string, versions []string, socketPath string) error
+
+// Watcher is the plugin watcher
+type Watcher struct {
+	path     string
+	handlers map[string]RegisterCallbackFn
+	stopCh   chan interface{}
+	fs       utilfs.Filesystem
+	watcher  *fsnotify.Watcher
+}
+
+// NewWatcher provides a new watcher
+func NewWatcher(sockDir string) Watcher {
+	return Watcher{
+		path:     sockDir,
+		handlers: make(map[string]RegisterCallbackFn),
+		fs:       &utilfs.DefaultFs{},
+	}
+}
+
+// AddHandler registers a callback to be invoked for a particular type of plugin
+func (w *Watcher) AddHandler(handlerType string, handlerCbkFn RegisterCallbackFn) {
+	w.handlers[handlerType] = handlerCbkFn
+}
+
+// Creates the plugin directory, if it doesn't already exist.
+func (w *Watcher) createPluginDir() error {
+	if _, err := w.fs.Stat(w.path); os.IsNotExist(err) {
+		glog.Infof("Plugin directory at %s does not exist. Recreating.", w.path)
+		err := w.fs.MkdirAll(w.path, 0755)
+		if err != nil {
+			return fmt.Errorf("error (re-)creating driver directory: %s", err)
+		}
+	}
+	return nil
+}
+
+// Walks through the plugin directory to discover any existing plugin sockets.
+func (w *Watcher) traversePluginDir() error {
+	files, err := w.fs.ReadDir(w.path)
+	if err != nil {
+		return fmt.Errorf("error reading the plugin directory: %v", err)
+	}
+	for _, f := range files {
+		// Currently only supports flat fs namespace under the plugin directory.
+		// TODO: adds support for hierachical fs namespace.
+		if !f.IsDir() && filepath.Base(f.Name())[0] != '.' {
+			w.registerPlugin(path.Join(w.path, f.Name()))
+		}
+	}
+	return nil
+}
+
+func (w *Watcher) init() error {
+	if err := w.createPluginDir(); err != nil {
+		return err
+	}
+	return w.traversePluginDir()
+}
+
+func (w *Watcher) registerPlugin(socketPath string) error {
+	//TODO: Implement rate limiting to mitigate any DOS kind of attacks.
+	glog.V(2).Infof("registerPlugin called for socketPath: %s", socketPath)
+	client, conn, err := dial(socketPath)
+	if err != nil {
+		glog.Warningf("Dial failed at socket %s, err: %v", socketPath, err)
+		return err
+	}
+	defer conn.Close()
+	infoResp, err := client.GetInfo(context.Background(), &watcherapi.InfoRequest{})
+	if err != nil {
+		glog.Errorf("Failed to get plugin info using RPC GetInfo at socket %s, err: %v", socketPath, err)
+		return err
+	}
+
+	resp := new(watcherapi.RegistrationStatus)
+	if handlerCbkFn, ok := w.handlers[infoResp.Type]; ok {
+		var versions []string
+		for _, version := range infoResp.SupportedVersions {
+			versions = append(versions, version)
+		}
+		// calls handler callback to verify registration request
+		if verifyErr := handlerCbkFn(infoResp.Name, versions, socketPath); verifyErr != nil {
```

Comment:
You have a race here. Plugins may be notified of Registration after the first CSI / Device Plugin specific call

Reply:
I dont think that is a race by definition. In registration failure case, there will no specific call to plugin and in registration success case too, plugin is supposed to be having specific service up prior to making its socket available registration, therefore i dont this this order really matters.

Reply:
In the plugin lifecycle it seems reasonable to expect plugin specific calls to be executed after the registration phase has finished.

Let's not have unexpected behaviors just because a plugin is supposed to have the services up prior to registering.

Reply:
It  can be handled simply by introducing an additional channel between watcher and each submodule but tbh that seems unnecessarily complicating the interface with no practical need. Why would a plugin advertise a socket until its server is not ready to serve.
Additionally, if majority feels it a "MUST" to handle, i am open to introduce a sync channel per handler to notify sub-modules status of NotifyRegistrationStatus.

Reply:
It would be simpler to defer the responsibility of notifying the registration status on the plugin module (i.e: device plugin or CIS)